ffyaml/ffyaml.go

```diff
@@ -48,29 +55,36 @@ func valsToStrs(val interface{}) ([]string, error) {
 	if err != nil {
 		return nil, err
 	}
-	return []string{s}, nil
+	return []*string{s}, nil
 
 }
 
-func valToStr(val interface{}) (string, error) {
+func valToStr(val interface{}) (*string, error) {
+	rtnStr := ""
+	var err error
+
 	switch v := val.(type) {
 	case byte:
-		return string([]byte{v}), nil
+		rtnStr = string([]byte{v})
 	case string:
-		return v, nil
+		rtnStr = v
 	case bool:
-		return strconv.FormatBool(v), nil
+		rtnStr = strconv.FormatBool(v)
 	case uint64:
-		return strconv.FormatUint(v, 10), nil
+		rtnStr = strconv.FormatUint(v, 10)
 	case int:
-		return strconv.Itoa(v), nil
+		rtnStr = strconv.Itoa(v)
 	case int64:
-		return strconv.FormatInt(v, 10), nil
+		rtnStr = strconv.FormatInt(v, 10)
 	case float64:
-		return strconv.FormatFloat(v, 'g', -1, 64), nil
+		rtnStr = strconv.FormatFloat(v, 'g', -1, 64)
+	case nil:
+		return nil, nil
```

Comment:
So if I'm reading this right, if the YAML parser yields a key with a nil value — presumably what happens when you just write `key: ` with nothing to the right of the colon — we would previously treat this as a StringConversionError, and are now special-casing it and treating it as if the `key` flag hadn't been provided at all. Is that correct?

Reply:
> and are now special-casing it and treating it as if the `key` flag hadn't been provided at all.

Well, the difference here is that in `valToStr()` (lines 81-82), in the case of a nil `value` for the value part of the YAML entry, `valToStr()` simply returns a `nil` (str pointer).

Then on line 29 of Parser(), we do a nil-check of the pointer for that value. Only if it is Not-Nil,  do we then go ahead and update the  `value` variable in the `set()` function.

Reply:
Yes, so we're changing 

```yaml
flagname: 
```

from being an error, to being a no-op, taking no action on this line at all, and specifically _not_ treating it as equivalent to `-flagname=""`. Is that correct?

Reply:
> from being an error, to being a no-op, taking no action on this line at all, and specifically not treating it as equivalent to -flagname="". Is that correct?


@peterbourgon. Yes, that is correct.

Reply:
@aramcodz Great! Thanks for confirming. So I agree with you that

```yaml
foo: 123
bar: 
log-level: debug
```

is valid YAML, which declares the key bar as having an empty/null value. And I think I'm convinced that the current behavior of the YAML parser, which treats this as an error condition, is incorrect and should be changed. But I'm not (yet) convinced that the behavior in this PR is what we should be doing. My intuition is that the above config should not be interpreted as bar not being passed at all, e.g.

```
myprogram -foo="123"         -log-level="debug"
```

but should rather be interpreted as bar being passed and given an empty value, e.g.

```
myprogram -foo="123" -bar="" -log-level="debug"
```

Do you have an opinion? If you prefer the first form, why? 🙃

Reply:
Yes, We prefer the first form (*Note: I've been working closely with my colleague, https://github.com/David-N-Perkins on this issue, and we conferred on the response).

We don't think that we can assume a blank YAML field value entry = an empty string value.  Let me explain:

1. Since the YAML specification treats a blank value as being a "null field" entry and not an empty string value, ("") we think that the ff library should respect that.
2.  _Not specifying a value_ is very different from specifying an empty value. A yaml config could have `bar: ""` to specify a blank (`string`) value.
3. Assuming that a blank YAML value equals `""` _also assumes a type_ (`string` type). We think `nil` is a much better representation of what was provided in the config _than assuming a type and a value._

Perhaps, ignoring a blank entry is not the best approach, but we see that as an issue with the existing ff lib design. A better approach might be **using pointers** instead of types for each config (FlagSet) value, so the user can differentiate between a config not being specified, specified without a value, or specified with a value. But we don't see how the current design allows for that.

Either way, we think _treating blanks as **not present**_ is better than assuming a blank YAML value is always an empty string.

Reply:
> Since the YAML specification treats a blank value as being a "null field" entry and not an empty string value, ("") we think that the ff library should respect that.

As far as I can see, "null field" is a concept defined by the YAML spec but without any specific semantics. When it needs to be interpreted in the context of a YAML AST being translated to flag values, it is not at all obvious to me what it should mean. 

Concretely, your PR makes

```yaml
foo: bar
baz:
```

equivalent to 

```yaml
foo: bar
```

I could be convinced that's the correct interpretation, but I don't believe it by default. Given `baz: `, it's much more intuitive to me that I get `-baz=""` out the other side, versus nothing at all. Why else would you put `baz: ` in the config in the first place? — I could be wrong! I don't know YAML super good, is there anything in the spec that would maybe push my belief one way or the other?

> Not specifying a value is very different from specifying an empty value. A yaml config could have bar: "" to specify a blank (string) value.

That's super confusing. To me it seems obvious that

```yaml
bar: ""
```

is, to a human, exactly the same as

```yaml
bar: 
```

In what [common] circumstances would someone leverage this null value property of YAML?

> Assuming that a blank YAML value equals "" also assumes a type (string type). We think nil is a much better representation of what was provided in the config than assuming a type and a value.

Given a flag `-key=val`, then yes, the val is by definition a string type. Maybe that's a disconnect here? All ff config files are by definition a map of strings to strings. If you want to specify a value as a non-string type, that's fine, but it will be converted to a string before it's received by ff.

>  A better approach might be using pointers instead of types for each config (FlagSet) value, so the user can differentiate between a config not being specified, specified without a value, or specified with a value. 

ff stands for flags-first, and it interprets everything, including parameters parsed from a config file, as if they were flags provided at the command line. The way to express "not specified" when enumerating flags at the command line is to not provide the flag. Therefore, intuitively, the way to express "not specified" when enumerating parameters in a config file is to not include the parameter in the file. 

Again — I'm open to being convinced otherwise. But show me something that will convince me! 🙃

Reply:
UPDATE.

@peterbourgon, As per your suggestion/argument in favor of blank strings, I modified the approach of my suggested YAML parser fix/change. It now will always return an empty string ("") for the "nil-case". 

Important: the implication is that `Parse()` will now return Parse errors if the blank YAML entry is for a config field that is of a type other than string or slice (of strings). See new/updated unit test function `TestParserReturnsErrorsForBlankNonStrings()`

Please see latest commits:
1. https://github.com/aramcodz/ff/commit/75e3485c94788728dddff61ab48e50436849294e
2. just fixed imports in ffyaml_test.go -> https://github.com/aramcodz/ff/commit/5e9383ff9ba0f3f4f0f09b5cc9d4fcd12aa8a14d

Is this solution to the YAML parser issue with blank fields acceptable?

Reply:
> The implication is that Parse() will now return Parse errors if the blank YAML entry is for a config field

Rather, a flag name.

> that is of a type other than string or slice (of strings). 

OK. Should it? 

Every flag value starts as a string and is converted to the flag's type. The JSON parser stringifies all of its values before passing them on. This will happen to the YAML values, too! What you're doing here is an extra layer of validation, which is then repeated, essentially, by the flag parsing code.

Does that all make sense? Do you still want to validate types at the parser level?

Reply:
> What you're doing now is an extra layer of type validation at the file parser level, and before the flag parsing code runs. The JSON parser, for example, doesn't do this — it stringifies all of its values before passing them on, so doesn't do this kind of thing.

@peterbourgon: I am really confused by your response. I did not add in **_any** type validation_ into the YAML parser.  Literally, to the actual implementation (non-test) ff code, the only change I am suggesting is adding in to `ValToStr()` function, lines 73-74:

```
	case nil:
		return "", nil

```

That 2-line addition is now the only proposed change in this PR to any implementation code in ff. 

And, in the test code, the most significant change I made is adding in new test function `TestParserReturnsErrorsForBlankNonStrings()`. This new test function demonstrates that it is indeed the `Parse()` function in `parse.go`, Not anything in `ffyaml.go` `Parser()`, that returns parse errors for blank YAML entries if the pre-defined Flagset type is something other than a string or slice of strings.  

So, I am really confused. 

Specifically, what is it in the implementation code you actually want me to change to fix the blank YAML entry error problem? 

What is it you will accept in a PR as a change to fix this issue with blank entries the YAML parser?

Reply:
Sorry, I should have clarified. I understand that the current ffyaml parser already does this validation, and that you haven't changed anything in that regard. I'm asking if it should _continue_ doing that validation. If your position is that your change doesn't affect the current behavior and so you don't care, then that's fine 👍

Reply:
>  If your position is that your change doesn't affect the current behavior and so you don't care, then that's fine.

I am focused on fixing the blank entries in YAML files issue. That's it. That is my priority because my team at IBM Watson Health has been developing some software and has been using **ff** as a dependency in our most recent update one of our services, specifically an overhaul to this (older, Preview Version) https://github.com/Alvearie/hri-mgmt-api. 

My PR really all about solving that problem (or, perhaps, at least changing the current behavior to be much more user friendly) with a small, incremental, least intrusive, change to the ff YAML parser. 

Making that one change in ff means it meets all our needs as a config library. That is, currently, this behavior (with the blank YAML entries) causes an issue for us during deployments that our users cannot understand exactly why the deployment failed. 

Hence, really all I am trying to accomplish here is a newly released version of ff, with the Blank-Entries-YAML-Files issue behavior "fixed", behaving in a way seems like a significant improvement for end-users over current behavior. 

We would _prefer_ contributing this proposed fix/improvement back the the **ff** OS project, as good Open Source community members, rather than just writing our own version of the parser itself.

Reply:
All good. Sorry for the distraction, then.